### ISSUE TITLE
[Feature request] Windows ARM support

### ISSUE THREAD

**Issue by TommyTran732** [mullvad/mullvadvpn-app#6102]
### I have checked if others have suggested this already

- [X] I have checked this issue tracker to see if others have reported similar issues.

### Feature description

It would be great to have Mullvad functioning on Windows ARM. Right now I only get this after installing Mullvad:

<img width="338" alt="Screenshot 2024-04-09 at 20 44 23" src="https://github.com/mullvad/mullvadvpn-app/assets/57488583/1092c6b3-20cb-426a-a0d1-65b251636d95">

### Alternative solutions

None

### Type of feature

- [ ] Better privacy/anonymity
- [ ] Better at circumventing censorship
- [ ] Easier to use
- [X] Other

### Operating System

- [ ] Android
- [ ] iOS
- [X] Windows
- [ ] macOS
- [ ] Linux

**Comment by dpaoliello** [comment]
I've started work on this, it looks like it will need a few changes - x64 emulation on ARM64 will not work since Mullvad includes a kernel driver.

Issues:
- [X] `maybenot` uses an old version of `ring` which doesn't support ARM64: <https://github.com/maybenot-io/maybenot/pull/23>
- [ ] `libwfp` is missing an ARM64 build config: <https://github.com/mullvad/libwfp/pull/39>
- [ ] `windows-libraries` is missing an ARM64 build config: <https://github.com/mullvad/windows-libraries/pull/46>
- [ ] `win-split-tunnel` has an x64-only build.bat and a broken ARM64 build: <https://github.com/mullvad/win-split-tunnel/pull/38>
- [ ] `apisocks5` needs to support building as ARM64: <https://github.com/mullvad/apisocks5/pull/1>
- [ ] `wireguard-nt` needs to support building as ARM64: WIP
- [ ] `dist-assets/binaries` needs to add support for ARM64: WIP
- [ ] `dist-assets/binaries` needs a copy of the `aarch64-pc-windows-msvc` binaries: TODO (will need help for this)
- [ ] `mullvadvpn-app` needs to add support for ARM64: WIP

**Comment by 0bscenity** [comment]
> Would really appreciate an ARM version for Windows 11 please.

agreed, openvpn config doesnt seem to work so rn i have no alternative

**Comment by dpaoliello** [comment]
> > Would really appreciate an ARM version for Windows 11 please.
> 
> agreed, openvpn config doesnt seem to work so rn i have no alternative

WireGuard has an ARM64 Windows app, and that's been working great for me with Mullvad's configs

**Comment by dpaoliello** [comment]
> > WireGuard has an ARM64 Windows app, and that's been working great for me with Mullvad's configs
> 
> Do you have a guide to set this up please?

WireGuard Windows apps are available at <https://download.wireguard.com/windows-client/>

Mullvad has a guide on how to generate a config file to load into the app: <https://mullvad.net/en/help/wireguard-app-windows>

**Comment by bakahk** [comment]
**1)** Download and install: https://download.wireguard.com/windows-client/ => wireguard-arm64-0.5.3.msi
**2)** Login: https://mullvad.net/account/wireguard-config?platform=windows => MULLVAD ID => WireGuard configuration file generator…
**3)** Generate key => Select one or multiple exit locations => Download zip archive…
_*You can generate and download archives for "several countries"._
**4)** Open the WireGuard app => Click Add Tunnel, select the downloaded file, then click Open.
_*That's it, you can connect to selected VPN servers._

It's not a pretty solution, much less user-friendly than the MullvadVPN app, but it works.

**Comment by 0bscenity** [comment]
> > > Would really appreciate an ARM version for Windows 11 please.
> > 
> > 
> > agreed, openvpn config doesnt seem to work so rn i have no alternative
> 
> WireGuard has an ARM64 Windows app, and that's been working great for me with Mullvad's configs

this is what i figured out, i dont link wireguard and prefer openvpn. I havent used wg since its infancy and it has gotten much better tho so using it seems much better know


> **1)** Download and install: https://download.wireguard.com/windows-client/ => wireguard-arm64-0.5.3.msi **2)** Login: https://mullvad.net/account/wireguard-config?platform=windows => MULLVAD ID => WireGuard configuration file generator… **3)** Generate key => Select one or multiple exit locations => Download zip archive… _*You can generate and download archives for "several countries"._ **4)** Open the WireGuard app => Click Add Tunnel, select the downloaded file, then click Open. _*That's it, you can connect to selected VPN servers._
> 
> It's not a pretty solution, much less user-friendly than the MullvadVPN app, but it works. _*Tested on Lenovo YOGA Slim 7x_

works on samsung galaxy books s

**Comment by dpaoliello** [comment]
I have this working locally now:
![image](https://github.com/user-attachments/assets/928b1c3b-6b67-4580-99ad-2ed1f5961be3)

PR has been updated with the changes.

**Comment by dpaoliello** [comment]
Correction: WireGuard is working, OpenVPN is not.

**Comment by TommyTran732** [comment]
Awesome! Even with just WireGuard it would still be great to just have it be available

**Comment by dpaoliello** [comment]
Everything has been merged in, now we just need a beta release!
Many thanks to @dlon for guiding me through this, making fixes, building binaries and doing reviews.
We also got OpenVPN working, but it is using the x64 exe.

**Comment by aelharda** [comment]
@dpaoliello Where can we find the beta installer please?

**Comment by dpaoliello** [comment]
> Where can we find the beta installer please?

Sorry, I'm not on the Mullvad team, so I have no control over when installers they release.

@dlon do you know when we might see a beta installer for Windows ARM64?

**Comment by talynone** [comment]
Any news on beta release?

**Comment by hulthe** [comment]
We don't support WIndows on ARM (_yet..._), we make **no guarantees** that it will work, it might void your warranty, etc etc.

That being said, we have made an ARM build for the latest beta, you can get the exe here:
https://releases.mullvad.net/desktop/releases/2024.6-beta1/MullvadVPN-2024.6-beta1_arm64.exe

And the signature here:
https://releases.mullvad.net/desktop/releases/2024.6-beta1/MullvadVPN-2024.6-beta1_arm64.exe.asc
https://mullvad.net/en/help/verifying-signatures

Our goal is to officially support Windows on ARM, but we're not quite there yet :)
If you do give it a try, let us know if it works! (or if ends up exploding your PC...)

(ping @talynone @aelharda)

**Comment by talynone** [comment]
@hulthe Seems to work fine on my Lenovo Slim 7x Windows 11 ARM computer, ran the Mullvad leak tests and everything is green, thanks!

**Comment by dpaoliello** [comment]
+1, working on my Surface Pro 11
(nice to be using a pre-built version instead of the one I build myself...)

**Comment by wminner** [comment]
Working with my Dell XPS 13 (9345 model).  Thanks!

**Comment by HyCore** [comment]
working also here on my lenovo thinkpad T14S.
Butter smooth. Well done !

**Comment by jasonscottcarter** [comment]
Works on Surface Laptop.  I get notification there is a new update to the Beta in the app, but don't upgrade guys, it lacks ARM support so I had to reinstall this ARM beta.

**Comment by orvitpng** [comment]
Just for the sake of documenting this, there is also a build for the latest beta: https://releases.mullvad.net/desktop/releases/2024.6-beta2/MullvadVPN-2024.6-beta2_arm64.exe.

**Comment by jasonscottcarter** [comment]
Great, just installed Beta 2.  The notification about there being a new version of the Beta has gone away now.  Is the ARM beta on it's own channel now?

**Comment by kvnlpz** [comment]
Not working on Snapdragon(R) X 12-core X1E80100

**Comment by jasonscottcarter** [comment]
@kvnlpz what model/brand of laptop?  Is that the one newer Snapdragon that is only on the Samsung Galaxy Book?

**Comment by dpaoliello** [comment]
> Not working on Snapdragon(R) X 12-core X1E80100

Can you also please describe what is not working, or what errors that you're seeing?

Are you using WireGuard or OpenVPN?

What version of Windows are you running?

**Comment by kvnlpz** [comment]
I uninstalled the VPN client and reinstalled it to try to reproduce the issue but this time it worked flawlessly. The first time I installed it, it lagged for a bit. Then when running the VPN through the tray icon, it would hang and not open up for around 5 minutes. When it finally popped up, it immediately said something about not being able to connect, even though I had not used it or signed into it yet. There has not been any updates to my system as far as I'm aware (no automatic updates). I am using Windows on Arm, Windows 11 Home, Version: 24H2. It's a Microsoft Surface Laptop

**Comment by jasonscottcarter** [comment]
Another one here:  https://releases.mullvad.net/desktop/releases/2024.7/  - installed and worked fine on Surface Laptop

**Comment by jasonscottcarter** [comment]
Just upgraded to the 2024.8 release for ARM.  Seems really stable.  When is there going to be an official ARM support release?

**Comment by Serock3** [comment]
@jasonscottcarter We are currently working on setting up our automated tests for the Windows ARM builds, and on some other minor hurdles. The release shouldn't be too far off, hopefully! 🪟 🦾

**Comment by dan-c-1488** [comment]
> [@jasonscottcarter](https://github.com/jasonscottcarter) We are currently working on setting up our automated tests for the Windows ARM builds, and on some other minor hurdles. The release shouldn't be too far off, hopefully! 🪟 🦾

Any news on the ARM build?

**Comment by Serock3** [comment]
Official Windows ARM support will come with the next release, and is already available in the  [`2025.3-beta1`](https://mullvad.net/en/download/vpn/windows#beta)! 🎉

**Comment by dpaoliello** [comment]
Is there a place to file issues for the website? Looks like it's still offering the x64 version for Arm64 devices.

Microsoft docs on how to detect arch via Javascript: https://learn.microsoft.com/en-us/microsoft-edge/web-platform/how-to-detect-win11#sample-code-for-detecting-arm-or-x86
